barretenberg/cpp/pil/vm2/nullifier_check.pil

```diff
@@ -0,0 +1,175 @@
+include "merkle_check.pil";
+include "ff_gt.pil";
+include "poseidon2_hash.pil";
+include "constants_gen.pil";
+include "precomputed.pil";
+
+/**
+ * This gadget checks reads and writes in the nullifier tree. The nullifier tree is an indexed tree where leaves
+ * contain nullifiers. When we read from the nullifier tree, we check whether a nullifier exists or not.
+ * For this we perform a low leaf membership proof and:
+ *  - if the low leaf nullifier is equal to the nullifier, that means that nullifier is present in the tree, and we assert that
+ *    exists is true.
+ *  - if the low leaf nullifier is not equal to the nullifier, we assert that the low leaf is indeed a valid low leaf for the
+ *    requested nullifier, proving non existence of the nullifier in the tree. In that case we check exists is false.
+ * In order to validate that a leaf is a low leaf of the nullifier, we need to check that the low_leaf.nullifier is < nullifier
+ * and that low_leaf.next_nullifier is > nullifier. However, we need to consider the case where next_nullifier is zero, which
+ * means "infinity". The highest nullifier inserted in the tree will point to infinity as the "next_nullifier".
+ * In order to write, we perform the above, and also:
```

Comment:
One nit that could be added to the above (IIRC) : We assume the newly inserted nullifier will never be the smallest value in the tree, as the tree is pre-filled with at least a single element prior.